src/nlp/meta.jl

```diff
@@ -161,25 +162,27 @@ struct NLPModelMeta <: AbstractNLPModelMeta
     @lencheck nnnet nnet
     @lencheck nlnet lnet
     @rangecheck 1 ncon lin nln nnet lnet
+    S = typeof(x0)
+    @assert S == typeof(lvar) == typeof(uvar) == typeof(y0) == typeof(lcon) == typeof(ucon)
```

Comment:
It would be better to set `x::S`, `lvar::S`, etc., and use `T = eltype(S)`, but I understood it's not possible to define the starting values for that case.

I wonder if we should define this constructor as `NLPModelaMeta{T, S}` and define a constructor 
```
NLPModelMeta(n, x0::S; kwargs...) = NLPModelMeta{eltype(S), S}(n, x0=x0; kwargs...)
```
This would certainly break more, but it would be clearer. @dpo, what do you think?